### ISSUE TITLE
Autograd functionality doesn't pass PyTorch autograd check

### ISSUE THREAD

**Issue by riceric22** [locuslab/pytorch_fft#4]
```
import torch
import pytorch_fft.fft.autograd as afft
from torch.autograd import gradcheck, Variable

in = (Variable(torch.randn(5,10).double().cuda(), requires_grad=True),)*2
test = gradcheck(afft.Fft(), in)
print(test)
```

The above snippet prints false. 

@rotmanmi I think the autograd implementation may be missing a few well-placed transposes, though I'm not sure what exactly would be the right fix.

**Comment by kingkongkai** [comment]
The backpropagation has been implemented incorrectly, at least for the general case. The error is related to how Wirtinger-calculus/CR-calculus works. It is true that the gradient of a Fourier transform function F(W) wrt. the input W is given by the inverse Fourier transform of the complex conjugate of the input gradient G: grad(F) = F^{-1}(G*), but you're not supposed to use the conjugation until you're at the layer where you need to update the network parameters (otherwise you would end up conjugating the gradients between every layer in a complex valued network, which makes no mathematical sense). So unless the gradients are used for updating network parameters that are fed directly into the Fourier transform, without any intermediate computations, your implementation will fail. If this is not the case, then the correct form of the gradient is actually F(G), and similarly for the inverse function F^{-1}(W), the gradient should be F^{-1}(G).

It still might not pass the gradcheck test though, as gradcheck uses differentiation by finite differences to control for errors, and I doubt they have implemented this with complex numbers in mind (not even sure if finite differences exists for functions of complex arguments). However, since the Fourier transform is a holomorphic function, it is also complex differentiable and should behave similarly to the regular calculus we are used to. So maybe it will give a True if you change the gradients like I said, but in any case I would avoid relying on pytorch's gradcheck when testing gradients of functions with complex arguments or complex values. I suggest using simple single-data-point regression networks instead, to see if it converges.

**Comment by riceric22** [comment]
Fixed via #7
